### PR TITLE
ASG module draft version. No output yet.

### PR DESCRIPTION
**Pull request by schweppes2205**

### CODE REVIEW

File: task_5/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Working with AWS ASG and AWS ECS services
+
+1. Need to create a module for ASG (includes asg and launch configuration)
+2. Need to create a module for ECS specific resources (ECS service, ECS cluster, ECS task definition)
+3. Add user data into ASG module to include EC2 instances into ECS cluster.
+4. Deploy centos container (https://hub.docker.com/_/centos) to ECS using your modules
+
+# Questions:
+
+- How to correctly organize the launch config change? Current solution causes tf apply failure after launch config change. In case launch config is changed and needed to be recreated but that is blocked by ASG that hat existing launch config attached. So somehow i need to create a new launch config, attach it to ASG and only after that drop the old one. The lifecycle is not helpful because it creates the launch config with the existing name.
```
Comment (alsar1980) [inline task_5/README.md:10]: Please look at https://github.com/palantir/bouncer 
Anyway this is not an issue for this task

File: task_5/module/sg/variables.tf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  type        = map
+  default = {
+    First_name = "firstName"
+    Last_name  = "secondName"
+  }
+}
+
+variable "sgr_ingress" {
+  description = "security group rules - ingress"
+  type        = list(map(string))
```
Comment (alsar1980) [inline task_5/module/sg/variables.tf]: Port numbers are not strings. It is better to use another type for this var

**Review by alsar1980 — COMMENTED** [review]
No outputs for ECS module

File: task_5/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+- Noticed that in current configuration in case of changes on any module the plan or apply mechanism detects changes on launch config with no reason (ami data source shows some changes)
```
Comment (alsar1980) [inline task_5/README.md:11]: Try to move data source to the root module

File: task_5/module/sg/variables.tf (hunk 2 above)
Comment (schweppes2205) [reply task_5/module/sg/variables.tf]: list of maps with string inside
```terraform
type = list(map(string))
```
is changed to custom object
```terraform
type = list(object({
    from_port   = number
    to_port     = number
    protocol    = string
    description = string
  }))
```
new default value example
```terraform
default = [{
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    description = "SSH"
    },]
```

File: task_5/README.md (hunk 3 above)
Comment (schweppes2205) [reply task_5/README.md:11]: done in last commit. the issue mentioned in the second note is resolved.

**Pull request merged**